staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+					"e": withDefault("x", stringType),
+				}),
+				"a1": *objectTypePtr(map[string]schema.Structural{
+					"b1": *objectTypePtr(map[string]schema.Structural{
+						"c1": integerType,
+					}),
+					"d2": *objectTypePtr(map[string]schema.Structural{
+						"e2": integerType,
+					}),
+				}),
+				"a3": *objectTypePtr(map[string]schema.Structural{
+					"b3": withDefault(map[string]interface{}{"c3": "y"}, *objectTypePtr(map[string]schema.Structural{
+						"c3": stringType,
+					})),
+				}),
+			}),
+			// https://github.com/google/cel-spec/blob/master/doc/langdef.md#field-selection
+			valid: []string{
+				"has(a.b)",
+				"!has(a.c)",
+				"has(a.d)",
+				"has(a.e)",
+				"has(a1.b1.c1)",
+				"!(has(a1.d2) && has(a1.d2.e2))", // must check intermediate optional fields (see below no such key error for d2)
+				"!has(a1.d2)",
+			},
+			errors: map[string]string{
+				"has(a.z)":      "undefined field 'z'",             // may not reference undefined fields, not even with has
+				"a['b'] == 1":   "no matching overload for '_[_]'", // only allowed on maps, not objects
+				"has(a1.d2.e2)": "no such key: d2",                 // has only checks last element in path, when d2 is absent in value, this is an error
+			},
+		},
+		{name: "map access",
+			obj: map[string]interface{}{
+				"val": map[string]interface{}{
+					"b": 1,
+					"d": 2,
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"val": mapType(&integerType),
+			}),
+			valid: []string{
+				"!has(val.a)",
+				"has(val.b)",
+				"!has(val.c)",
+				"has(val.d)",
+				"val.all(k, val[k] > 0)",
+				"val.exists(k, val[k] > 1)",
+				"val.exists_one(k, val[k] == 2)",
+				"!val.exists(k, val[k] > 2)",
+				"!val.exists_one(k, val[k] > 0)",
+				"size(val) == 2",
+				"val.map(k, val[k]).exists(v, v == 1)",
+				"size(val.filter(k, val[k] > 1)) == 1",
+			},
+			errors: map[string]string{
+				"val['c'] == 1": "no such key: c",
+			},
+		},
+		{name: "listMap access",
+			obj: map[string]interface{}{
+				"listMap": []interface{}{
+					map[string]interface{}{"k": "a1", "v": "b1"},
+					map[string]interface{}{"k": "a2", "v": "b2"},
+					map[string]interface{}{"k": "a3", "v": "b3", "v2": "z"},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"listMap": listMapType([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k":  stringType,
+					"v":  stringType,
+					"v2": stringType,
+				})),
+			}),
+			valid: []string{
+				"has(listMap[0].v)",
+				"listMap.all(m, m.k.startsWith('a'))",
+				"listMap.all(m, !has(m.v2) || m.v2 == 'z')",
+				"listMap.exists(m, m.k.endsWith('1'))",
+				"listMap.exists_one(m, m.k == 'a3')",
+				"!listMap.all(m, m.k.endsWith('1'))",
+				"!listMap.exists(m, m.v == 'x')",
+				"!listMap.exists_one(m, m.k.startsWith('a'))",
+				"size(listMap.filter(m, m.k == 'a1')) == 1",
+				"listMap.exists(m, m.k == 'a1' && m.v == 'b1')",
+				"listMap.map(m, m.v).exists(v, v == 'b1')",
+			},
```

Comment:
What if you where to do:
```
listMap.exists(m, m.v2 == 'z')
```
I believe this will result in a `"no such key: v2"` error.

I think it might be worth testing this case where one of the field of the object doesn't exist in all the elements of the list.

Reply:
You found something interesting I didn't know about. It's hinted at in the fine print of the [CEL spec about macros](https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros):

> The all() macro combines per-element predicate results with the "and" (&&) operator, so if any predicate evaluates to false, the macro evaluates to false, ignoring any errors from other predicates.

Here's the observed behavior:

```
listMap.exists(m, m.v2 == 'z') -> true
listMap.all(m, m.v2 != 'z') -> false

listMap.all(m, m.v2 == 'z') -> no such key: v2
// exists one() is stricter than map() or exists(), it requires exactly one predicate evaluate to true and the rest
listMap.exists_one(m, m.v2 == 'z') -> no such key: v2
listMap.filter(m, m.v2 == 'z') -> no such key: v2
listMap.map(m, m.v2) -> no such key: v2
```
I've expanded the tests to cover all of these plus some commentary and some examples of how to use has() to avoid the no such key errors.